akka-stream/src/main/scala/akka/stream/impl/io/OutputStreamSourceStage.scala

```diff
@@ -41,119 +39,57 @@ final private[stream] class OutputStreamSourceStage(writeTimeout: FiniteDuration
 
     require(maxBuffer > 0, "Buffer size must be greater than 0")
 
-    val dataQueue = new LinkedBlockingQueue[ByteString](maxBuffer)
+    // Semaphore counting the number of elements we are ready to accept,
+    // which is the demand plus the size of the buffer.
+    val semaphore = new Semaphore(maxBuffer, /* fair =*/ true)
```

Comment:
What buffer?

Reply:
we accept `maxBuffer` 'write' actions before blocking and waiting for downstream demand to make more room. The buffer is not very visible because we use `emit` which does the buffering when there is no demand yet.

Reply:
Ah, I see. push vs emit.

Reply:
Maybe mention that in the comment, just if there is some other moron like me who doesn't get it reading the sources.